pkg/k8sops/generators.go

```diff
@@ -0,0 +1,392 @@
+				},
+				Ports:     ports,
+				ClusterIP: "None",
+			},
+		}, nil
+	case M3coordinatorSvcName:
+		ports, err := k.GenerateServicePorts(name)
+		if err != nil {
+			return nil, err
+		}
+		return &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: M3coordinatorSvcName,
+				Labels: map[string]string{
+					"app": M3coordinatorSvcName,
+				},
+			},
+			Spec: v1.ServiceSpec{
+				Selector: map[string]string{
+					"app": M3dbNodeSvcName,
+				},
+				Ports: ports,
+			},
+		}, nil
+	default:
+		return nil, errors.New("service is not defined")
+	}
+}
+
+// GenerateServicePorts generates the service ports object for m3db
+func (k *K8sops) GenerateServicePorts(name string) ([]v1.ServicePort, error) {
+	switch name {
+	case M3dbNodeSvcName:
+		return []v1.ServicePort{
+			v1.ServicePort{
+				Name:     "client",
+				Port:     9000,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "cluster",
+				Port:     9001,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "http-node",
+				Port:     9002,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "http-cluster",
+				Port:     9003,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "debug",
+				Port:     9004,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "query",
+				Port:     7201,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "query-metrics",
+				Port:     7203,
+				Protocol: v1.ProtocolTCP,
+			},
+		}, nil
+	case M3coordinatorSvcName:
+		return []v1.ServicePort{
+			v1.ServicePort{
+				Name:     "query",
+				Port:     7201,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "query-metrics",
+				Port:     7203,
+				Protocol: v1.ProtocolTCP,
+			},
+		}, nil
+	default:
+		return nil, errors.New("serivce is not defined")
+	}
+}
+
+const (
+	_probeTimeoutSeconds      = 30
+	_probeInitialDelaySeconds = 10
+	_probeFailureThreshold    = 15
+	_probePort                = 7201
+	_probePath                = "/health"
+
+	_baseImage              = "quay.io/m3/m3dbnode:latest"
+	_dataDirectory          = "/var/lib/m3db/"
+	_configurationDirectory = "/etc/m3db/"
+	_configurationName      = "m3dbnode-configuration"
+	_configurationFileName  = "m3dbnode.yml"
+)
+
+var (
+	_configurationFileLocation = fmt.Sprintf("%s%s", _configurationDirectory, _configurationFileName)
+)
+
+// GenerateStatefulSet provides a statefulset object for a m3db cluster
+func (k *K8sops) GenerateStatefulSet(
+	cluster *myspec.M3DBCluster,
+	clusterSpec myspec.ClusterSpec,
+	isolationGroup string,
+	instanceAmount *int32,
+) (*appsv1.StatefulSet, error) {
+	clusterName := cluster.GetName()
+	ssName := k.StatefulSetName(clusterName, isolationGroup)
+	ports, err := k.GenerateContainerPorts(M3dbNodeSvcName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse CPU / Memory
+	limitCPU, _ := resource.ParseQuantity(clusterSpec.Resources.Limits.CPU)
```

Comment:
nit: error handling for all of these Parse calls

Reply:
Done